src/main/java/appeng/items/tools/powered/powersink/AEBasePoweredItem.java

```diff
@@ -19,12 +19,194 @@
+	}
+
+	@Override
+	public boolean isDamageable()
+	{
+		return true;
+	}
+
+	@Override
+	protected void getCheckedSubItems( final CreativeTabs creativeTab, final NonNullList<ItemStack> itemStacks )
+	{
+		super.getCheckedSubItems( creativeTab, itemStacks );
+
+		final ItemStack charged = new ItemStack( this, 1 );
+		final NBTTagCompound tag = Platform.openNbtData( charged );
+		tag.setDouble( CURRENT_POWER_NBT_KEY, this.getAEMaxPower( charged ) );
+		tag.setDouble( MAX_POWER_NBT_KEY, this.getAEMaxPower( charged ) );
+
+		itemStacks.add( charged );
+	}
+
+	@Override
+	public boolean isRepairable()
+	{
+		return false;
+	}
+
+	@Override
+	public double getDurabilityForDisplay( final ItemStack is )
+	{
+		return 1 - this.getAECurrentPower( is ) / this.getAEMaxPower( is );
+	}
+
+	@Override
+	public boolean isDamaged( final ItemStack stack )
+	{
+		return true;
+	}
+
+	@Override
+	public void setDamage( final ItemStack stack, final int damage )
+	{
+
+	}
+
+	@Override
+	public double injectAEPower( final ItemStack is, final double amount, Actionable mode )
+	{
+		if( mode == Actionable.SIMULATE )
+		{
+			final double required = this.getAEMaxPower( is ) - this.getAECurrentPower( is );
+			final double overflow = amount - required;
+
+			return Math.max( 0, overflow );
+		}
+
+		return this.getInternalBattery( is, batteryOperation.INJECT, amount );
+	}
+
+	@Override
+	public double extractAEPower( final ItemStack is, final double amount, Actionable mode )
+	{
+		if( mode == Actionable.SIMULATE )
+		{
+			final double stored = this.getAECurrentPower( is );
+
+			return Math.min( amount, stored );
+		}
+
+		return this.getInternalBattery( is, batteryOperation.EXTRACT, amount );
+	}
+
+	@Override
+	public double getAEMaxPower( final ItemStack is )
+	{
+		return this.powerCapacity;
+	}
+
+	@Override
+	public double getAECurrentPower( final ItemStack is )
+	{
+		return this.getInternalBattery( is, batteryOperation.STORAGE, 0 );
+	}
+
+	@Override
+	public AccessRestriction getPowerFlow( final ItemStack is )
+	{
+		return AccessRestriction.WRITE;
+	}
+
+	@Override
+	public ICapabilityProvider initCapabilities( ItemStack stack, NBTTagCompound nbt )
+	{
+		return new PoweredItemCapabilities( stack, this );
+	}
+
+	private double getInternalBattery( final ItemStack is, final batteryOperation op, final double adjustment )
+	{
+		final NBTTagCompound data = Platform.openNbtData( is );
+
+		final double maxStorage = this.getAEMaxPower( is );
+		double currentStorage = data.getDouble( CURRENT_POWER_NBT_KEY );
+
+		switch( op )
+		{
+			case INJECT:
+				currentStorage += adjustment;
+				if( currentStorage > maxStorage )
+				{
+					final double diff = currentStorage - maxStorage;
+					data.setDouble( CURRENT_POWER_NBT_KEY, maxStorage );
+					return diff;
+				}
+				data.setDouble( CURRENT_POWER_NBT_KEY, currentStorage );
+				return 0;
+			case EXTRACT:
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1151)